### PR TITLE
Developer docs for frontend

### PR DESCRIPTION
**Pull request by ongspxm**
```
The new dashboard doesn't come with the updated development
documentation. And this can prove troublesome for new developers as
they will have to spend a lot of time in the current codebase to make
sense of how the dashboard is structured.

Let's include a short writeup of the dashboard framework to help
explain how the data is being rendered more clearly.
```

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -99,28 +99,43 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
     - `ReportGenerator` to determine the directory to output the report.
 
 
 ## HTML Dashboard
+The source files for the dashboard is located in `frontend/src` and is built by [spuild](https://github.com/ongspxm/spuild2) before being packaged into the JAR file to be extracted as part of the report.
 
- ![Dashboard](images/dashboard-architecture.png)
+### Javascript Files
+- **api.js** - loading and parsing of the dashboard content
+- **main.js** - main controller that starts the loading of the dashboard and handles the loading of the content into different modules
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: I think this should be sufficient?

main controller that ~starts the loading of the dashboard and~ handles the loading of the content into different modules

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -99,28 +99,43 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
  These configuration information are used by:
     - `GitDownloader` to determine which repository to download from and which branch to check out to.
     - `AuthorshipReporter` and `CommitsReporter` to determine the range of commits and files to analyze.
     - `ReportGenerator` to determine the directory to output the report.
 
 
 ## HTML Dashboard
+The source files for the dashboard is located in `frontend/src` and is built by [spuild](https://github.com/ongspxm/spuild2) before being packaged into the JAR file to be extracted as part of the report.
 
- ![Dashboard](images/dashboard-architecture.png)
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: need to upload a new diagram

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -99,28 +99,57 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+`window.app` is broken down into two main parts, the summary view and the tabs view. The basic design is to have a summary view (with the ramp charts) as the main dashboard, and other modules display additional information in the tabbed interface which is displayed on the right.
 
-  * The dashboard template is a set of `HTML` + `CSS` + `Javascript`, which is located in `template/`. It is used to interpret the `JSON` information and display the dashboard on a web browser. <br>
-  Whenever *RepoSense* is built using *Gradle*, the current dashboard template's content will be zipped into `templateZip.zip` file for easier retrieval, and packaging into the generated `.jar` file.<br>
-  The dashboard template will then be unzipped to the target location along with the generated `JSON` data files in order to produce the dashboard whenever *RepoSense* generates a new report.
+**overall architecture**
+
+![overall architecture](images/dashboard-architecture.png)
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: Place this at the top of this section instead, so there's a better flow from an overall of the architecture first, then to the details of each component.
Update the section title to `Dashboard Architecture` as well.

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -99,28 +99,57 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+### Loading of dashboard information
+When the dashboard is first loaded, the main vue object tries to retreive the `summary.json` file in order to determine the right `commits.json` files to load into memory. `api.js` handles the loading of the file, and approriately gets the relevant file information, depending on whether the network files is available or a report archive have to be used.
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: `Vue`
same for the rest below.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -99,28 +99,57 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+Once the relevant `commit.json` files hash been loaded, all the repo information will be passed into `v_summary` to be loaded in the summary view as the releveant ramp charts.
+
+### Activating additional view modules
+Most activity or actions should happen within the module itself, but in the case where there is a need to control the tab view from the module, an event is emitted from the module to main vue object (`window.app`), which then handles the data receive and pass it along to the relevant modules.
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: from the module to **the** main **V**ue object ...
handles the data receive**d** and pass**es** it along ...

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -99,28 +99,61 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+- **main.js** - main controller that handles the loading of the content into different modules
+- **v_summary.js** - module that supports the ramp chart view
+- **v_authorship.js** - module that supports the authorship view
 
-  * The dashboard template is a set of `HTML` + `CSS` + `Javascript`, which is located in `template/`. It is used to interpret the `JSON` information and display the dashboard on a web browser. <br>
-  Whenever *RepoSense* is built using *Gradle*, the current dashboard template's content will be zipped into `templateZip.zip` file for easier retrieval, and packaging into the generated `.jar` file.<br>
-  The dashboard template will then be unzipped to the target location along with the generated `JSON` data files in order to produce the dashboard whenever *RepoSense* generates a new report.
+Our project follows the [AirBNB Javascript Style Guide](https://github.com/airbnb/javascript), the eslint configuration file is available at the root of the project. Please run a `npm run lint -- --fix frontend/src/**/*js` from the project root directory and fix all of the eslint errors before committing your code for final review.
```
[the remark this answers is not in the document]
Comment (eugenepeh) [reply docs/DeveloperGuide.md]: Maybe add a header, `Code Quality`?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -99,28 +99,61 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
     - `GitDownloader` to determine which repository to download from and which branch to check out to.
     - `AuthorshipReporter` and `CommitsReporter` to determine the range of commits and files to analyze.
     - `ReportGenerator` to determine the directory to output the report.
 
 
 ## HTML Dashboard
+The source files for the dashboard is located in [`frontend/src`](https://github.com/reposense/RepoSense/tree/master/frontend/src) and is built by [spuild](https://github.com/ongspxm/spuild2) before being packaged into the JAR file to be extracted as part of the report.
 
- ![Dashboard](images/dashboard-architecture.png)
+If needed, the main HTML file is generated from `frontend/src/index.jade`.
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: ~If needed, t~`T`he main HTML file is generated from `frontend/src/index.jade`.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -99,28 +99,61 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
-  The dashboard template will then be unzipped to the target location along with the generated `JSON` data files in order to produce the dashboard whenever *RepoSense* generates a new report.
+Our project follows the [AirBNB Javascript Style Guide](https://github.com/airbnb/javascript), the eslint configuration file is available at the root of the project. Please run a `npm run lint -- --fix frontend/src/**/*js` from the project root directory and fix all of the eslint errors before committing your code for final review.
+
+### Dashboard Architecture
+![dashboard architecture](images/dashboard-architecture.png)
+
+The main Vue object (`window.app`) is responsible for the loading of the dashboard (through `summary.json`). Its `repos` attribute is tied to the global `window.REPOS`, and is passed into the various other modules when the information is needed.
+
+`window.app` is broken down into two main parts, the summary view and the tabs view. The basic design is to have a summary view (with the ramp charts) as the main dashboard, and other modules display additional information in the tabbed interface which is displayed on the right.
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: `window.app` is broken down into two main parts
- the summary view
- and the tabs view

Comment (eugenepeh) [reply docs/DeveloperGuide.md]: Summary view act as the main dashboard which shows the varies calculations.
Tabbed interface is responsible for loading various modules such as authorship to display additional information.

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -99,28 +99,61 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+### Loading of dashboard information
+When the dashboard is first loaded, the main Vue object tries to retreive the `summary.json` file in order to determine the right `commits.json` files to load into memory. `api.js` handles the loading of the file, and approriately gets the relevant file information, depending on whether the network files is available or a report archive have to be used.
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: The main Vue object depends on the `summary.json` data to determine the right `commits.json` files to load into memory. This is handled by `api.js`. It loads the relevant file information, depending on whether the network files is available or a report archive have to be used.

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -99,28 +99,61 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+Once the relevant `commit.json` files hash been loaded, all the repo information will be passed into `v_summary` to be loaded in the summary view as the releveant ramp charts.
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: `hash` -> `has`

Comment (eugenepeh) [reply docs/DeveloperGuide.md]: `releveant` -> `relevant`

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -50,6 +57,9 @@ This project follows [oss-generic coding standards](https://oss-generic.github.i
 
 Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document to configure *Intellij* to check style-compliance as you write code.
 
+### Configuring the JavaScript coding style
+Our project follows the [Airbnb Javascript Style Guide](https://github.com/airbnb/javascript), the eslint configuration file is available at the root of the project. Please run a `npm run lint -- --fix frontend/src/**/*js` from the project root directory and fix all of the eslint errors before committing your code for final review.
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: Just remembered, users may not have installed `npm`, so might want to at least add a note and link to that installation for this.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -99,28 +109,92 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+Summary view act as the main dashboard which shows the various calculations. </br>
+Tabbed interface is responsible for loading various modules such as authorship to display additional information.
+
+### Javascript Files
+- [**main.js**](../frontend/src/static/js/main.js) - main controller that pushes content into different modules
+- [**api.js**](../frontend/src/static/js/api.js)- loading and parsing of the dashboard content
+- [**v_summary.js**](../frontend/src/static/js/main.js) - module that supports the ramp chart view
+- [**v_authorship.js**](../frontend/src/static/js/main.js) - module that supports the authorship view
+
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: Can add on a section on `json` files, briefly describe what each file contain.
Will make it easier to follow when you mention the use of those `json` files below

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -99,28 +109,92 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+An additional thing to note is the event handler should handle the switching of the tabs as well, activating the display of the rights tabs just that the relevant tabs are being displayed at all times.
+
+#### Hash link
+Other than the global main Vue object, another global variable we have is the `window.hashParams`. This object is reponsible for generating the relevant permalink for a specific view of the summary module for the dashboard.
+
+### Data loader (api.js)
+This is the module that is in charged of loading and parsing the data files generated as part of the report.
+
+#### Loading from ZIP file
+Due security design, most modern browsers do not allow web pages to obtain local files using the directory alone. As such, a ZIP archive of the report information will be produced alongside the report generation.
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: Due **to** security design, most modern browsers (e.g. Chrome) do not...

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -99,28 +109,92 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+This archive can be used in place of the network files to load the information into the dashboard when the dashboard HTML.
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: ...into the dashboard when the dashboard HTML??? What does this mean?

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -99,28 +109,92 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+As such, the API module will be handling all request for all the JSON data files. If the network file is not available, the files will be obtained from the zip archive provided.
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: Capital **ZIP**
No need for `As such` in the beginning as well, this sentence doesn't link to the previous right?

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -99,28 +109,92 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+#### Retrieving and parsing information
+After the JSON files are loaded from their respective sources, the data will be parsed as objects and included inside the global storage object, `window.REPOS`,  in the right format.
+
+For the basic skeleton of `window.REPOS`, refer to the generated `summary.json` file in the report for more details.
+
+### Summary View (v_summary.js)
+The `v_summary` module is in charged of loading the ramp charts from the corresponding `commits.json`.
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: module is in charge~~d~~ of..

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -99,28 +112,99 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+### JSON report files
+- **summary.json** - a list of all the repositories and their respective details
+- **projName/commits.json** - contains information of the users' commits information (e.g. line deletion, insertion, etc), grouped by date
+- **projName/authorship.json** - contains information from git blame, detailing the author of each line for all the processed files
+
+### Main (main.js)
+This contains the logic for main VueJS object, `window.app`, which is responsible for passing the necessary data into the relevant modules to be loaded.
+
+`v_summary` and `v_authorship` are components which will be embedded into dashboard and will render the corresponding content based on the data passed into it from the main `window.app`.
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: actually i mean a diagram under each module. the architecture is too vague.

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -99,28 +112,99 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+Summary view act as the main dashboard which shows the various calculations. </br>
+Tabbed interface is responsible for loading various modules such as authorship to display additional information.
+
+### Javascript Files
+- [**main.js**](../frontend/src/static/js/main.js) - main controller that pushes content into different modules
+- [**api.js**](../frontend/src/static/js/api.js)- loading and parsing of the dashboard content
+- [**v_summary.js**](../frontend/src/static/js/v_summary.js) - module that supports the ramp chart view
+- [**v_authorship.js**](../frontend/src/static/js/v_authorship.js) - module that supports the authorship view
+
+### JSON report files
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: Consistency, if other triple hex headers capitalize first letter, this should follow.

File: docs/DeveloperGuide.md (hunk 17 above)
Comment (ongspxm) [reply docs/DeveloperGuide.md]: more drawings? okay :cry:

Comment (eugenepeh) [reply docs/DeveloperGuide.md]: if you don't have powerpoint etc... can do paper drawing, we try to convert it to diagram for you.

**Comment by ongspxm** [comment]
the diagram for summary can? if can i do for the authorship also

**Comment by eugenepeh** [comment]
> the diagram for summary can? if can i do for the authorship also

Looks good, 
2. can indicate the variable name?
3. why is there a `?`?
5. the variable affected?
6. isn't ramp chart generated by main?

**Comment by ongspxm** [comment]
3. the summary is asking for the filter params from the hash.
6. nope, all the display itself is drawn within the modules, `main` contains the submodules

**Comment by eugenepeh** [comment]
>`3.` the summary is asking for the filter params from the hash.
`6.` nope, all the display itself is drawn within the modules, `main` contains the submodules

`3.` in that case use words like `get` or `request`. the `?` there is not very explanatory.
`6.` that way, maybe name the action as `Draw ramp charts`? Having a past participle there raise the question of who did it.

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -99,28 +112,103 @@ Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document t
...
+
+#### Loading of dashboard information
+The main Vue object depends on the `summary.json` data to determine the right `commits.json` files to load into memory. This is handled by `api.js` which loads the relevant file information from the network files if it is available, otherwise a report archive, `archive.zip`, have to be used.
+
+Once the relevant `commit.json` files are loaded, all the repo information will be passed into `v_summary` to be loaded in the summary view as the relevant ramp charts.
+
+#### Activating additional view modules
+Most activity or actions should happen within the module itself, but in the case where there is a need to control the tab view from the module, an event is emitted from the module to the main Vue object (`window.app`), which then handles the data received and passes it along to the relevant modules.
+
+An additional thing to note is the event handler should handle the switching of the tabs as well, activating the display of the rights tabs just that the relevant tabs are being displayed at all times.
```
Comment (eugenepeh) [inline docs/DeveloperGuide.md]: this feature doesnt seems to be there anymore

**Comment by ongspxm** [comment]
@eugenepeh i tried, pls help check if sd charts are correct

**Comment by eugenepeh** [comment]
General feedback:
For the SD, try to use the name of the functions/variables so that other developers can easily pinpoint the target area.

image
image

**Comment by damithc** [comment]
Guys, try to do some simple (not UML) and minimal diagrams or else we end up wasting lot of time on diagramming. The current ones are still not compliant with the notation we teach. e.g., the way the loop/branch conditions are shown. 
Even if we eventually get these in the correct format, we still have to maintain them when code changes in future, which will be some more work. Usually we use PowerPoint as the source so that it is easy to modify later.

**Comment by eugenepeh** [comment]
> Guys, try to do some simple (not UML) and minimal diagrams or else we end up wasting lot of time on diagramming.

@damithc , in that case, we can probably continue using SD and only illustrate the main activity? 
Basically just the main objects that were affected in main.js code and main operation in the sub-modules. This help to give a quick and general impression of their purpose.

We can leave out minor operations such as `split file into touched & untouched` etc so that we don't have to continuously maintain them.

**Comment by damithc** [comment]
> @damithc , in that case, we can probably continue using SD and only illustrate the main activity?

Sounds good.

**Comment by ongspxm** [comment]
yay after 20 days, its finally done, i hope :(

**Comment by damithc** [comment]
> @damithc do you think its acceptable?

Sure. Go ahead.

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -125,28 +138,101 @@ gradlew run -Dargs="-view output_path/reposense-report"
...
 
 
 ## HTML Dashboard
+The source files for the dashboard is located in [`frontend/src`](../frontend/src) and is built by [spuild](https://github.com/ongspxm/spuild2) before being packaged into the JAR file to be extracted as part of the report.
+
+The main HTML file is generated from [`frontend/src/index.jade`](../frontend/src/index.jade).
+
+[Vue](https://vuejs.org/v2/api/) (pronounced /vjuː/, like view) is a progressive framework for building user interfaces. It is heavily ultilized in the dashboard to dynamically update the information in the various views. (Style guide available [here](https://vuejs.org/v2/style-guide/)).
+
+![dashboard screenshot](images/dashboard.png)
```
Comment (yamidark) [inline docs/DeveloperGuide.md]: Cant seem to see the dashboard screenshot?
Did you forget to add the image from the `odg` diagram?

Comment (eugenepeh) [reply docs/DeveloperGuide.md]: @ongspxm for your attention

Comment (ongspxm) [reply docs/DeveloperGuide.md]: oh gawd, its still alive... hahhah, okay, gonna check it

**Pull request merged**